### DISCUSSION TITLE
PyQt5WebKit in linux dev environment?

### DISCUSSION THREAD

**Discussion by terminalnode** [qutebrowser/qutebrowser#5994] [Q&A]
Hi
I'm trying to get the dev environment setup, and I'm wondering if PyQt5.WebKit is supposed to be unavailable on Linux? I can't install it because `pip` says:
```
ERROR: Could not find a version that satisfies the requirement PyQt5WebKit
ERROR: No matching distribution found for PyQt5WebKit
```

The PyPi page for PyQt5WebKit says:
> Wheels for the GPL version for 32 Windows can be installed from PyPI:

so I take it these parts of the application are supposed not to work on Linux?

Additionally, is the `mypy` and `check-manifest` tests supposed to fail? `mypy` gives 9 errors in 5 files and `check-manifest` complains that `qutebrowser/html/doc/img/cheatsheet-{big,small}.png` are not in VCS (which seems accurate since `qutebrowser/html/doc` is gitignored).

**Comment by The-Compiler** [comment]
I don't see a `PyQt5WebKit` on PyPI ([404](https://pypi.org/project/PyQt5WebKit/)), so I'm not sure where you're quoting this from. QtWebKit certainly works on Linux, but you'd need to install it via your distribution's packages or build it from source (it's not available via PyPI). Probably not too relevant though, depending on what you're working on exactly.

> Additionally, is the `mypy` and `check-manifest` tests supposed to fail?

No - they [pass on CI](https://github.com/qutebrowser/qutebrowser/actions/runs/449162311) and also for me locally.

> `mypy` gives 9 errors in 5 files

Is your mypy up to date? How are you running it? What are those errors?

> and `check-manifest` complains that `qutebrowser/html/doc/img/cheatsheet-{big,small}.png` are not in VCS (which seems accurate since `qutebrowser/html/doc` is gitignored).

It parses the output of `git status --porcelain` which does account for `.gitignore` - what does that command say on your system? I'd expect an empty output.

**Reply by andrew222651** [reply]
Is there a way of disabling webkit testing? The end-to-end tests often fail if qtwebkit isn't installed.

Edit: looks like `--qute-bdd-webengine` as defined in `tests/conftest.py`

**Reply by The-Compiler** [reply]
Either that, or use an environment which uses a fixed PyQt version such as `py39-pyqt515` - that will then use PyQt from PyPI which only comes with QtWebEngine, and thus automatically run QtWebEngine tests only.

**Comment by terminalnode** [comment]
My mistake, the package I was referencing is just called [PyQtWebKit](https://pypi.org/project/PyQtWebKit/). Can't find anything about it in my repositories (on ArchLinux), but if it's not too relevant I might just skip it I suppose. App still runs fine.

> Is your mypy up to date? How are you running it? What are those errors?

So because it's such a big project I decided to try out PyCharm for setting this all up, using the virtualenv provided by the project. mypy version is 0.790, mypy-extensions is 0.4.3.

Errors are:
```
mypy run-test-pre: PYTHONHASHSEED='1847270991'
mypy run-test: commands[0] | /home/terminal/code/qutebrowser/.tox/mypy/bin/python -m mypy qutebrowser tests
qutebrowser/misc/guiprocess.py:166: error: Too many arguments for
"startDetached" of "QProcess"  [call-arg]
            ok, _pid = self._proc.startDetached(
                       ^
qutebrowser/misc/guiprocess.py:167: error: unused 'type: ignore' comment
                cmd, args, None)  # type: ignore[call-arg]
                ^
qutebrowser/keyinput/modeman.py:290: error: Non-overlapping container check
(element type: "KeyboardModifiers", container item type: "KeyboardModifier") 
[comparison-overlap]
            has_modifier = event.modifiers() not in [
                           ^
qutebrowser/keyinput/modeman.py:293: error: unused 'type: ignore' comment
            ]  # type: ignore[comparison-overlap]
            ^
qutebrowser/browser/webengine/webenginetab.py:269: error: No overload variant
of "FindFlags" matches argument type "int"  [call-overload]
            flags = QWebEnginePage.FindFlags(
                    ^
qutebrowser/browser/webengine/webenginetab.py:269: note: Possible overload variants:
qutebrowser/browser/webengine/webenginetab.py:269: note:     def FindFlags(self, f: Union[FindFlags, FindFlag]) -> FindFlags
qutebrowser/browser/webengine/webenginetab.py:269: note:     def FindFlags(self, a0: FindFlags) -> FindFlags
qutebrowser/browser/webengine/webenginetab.py:269: note:     <1 more non-matching overload not shown>
qutebrowser/browser/webengine/webenginetab.py:270: error: unused 'type: ignore'
comment
                int(self._flags))  # type: ignore[call-overload]
                ^
qutebrowser/browser/webkit/webkittab.py:183: error: No overload variant of
"FindFlags" matches argument type "int"  [call-overload]
            flags = QWebPage.FindFlags(
                    ^
qutebrowser/browser/webkit/webkittab.py:183: note: Possible overload variants:
qutebrowser/browser/webkit/webkittab.py:183: note:     def FindFlags(self, f: Union[FindFlags, FindFlag]) -> FindFlags
qutebrowser/browser/webkit/webkittab.py:183: note:     def FindFlags(self, a0: FindFlags) -> FindFlags
qutebrowser/browser/webkit/webkittab.py:183: note:     <1 more non-matching overload not shown>
qutebrowser/browser/webkit/webkittab.py:184: error: unused 'type: ignore'
comment
                int(self._flags))  # type: ignore[call-overload]
                ^
qutebrowser/components/misccommands.py:79: error: Argument 1 to
"setWindowFlags" of "QWidget" has incompatible type "int"; expected
"Union[WindowFlags, WindowType]"  [arg-type]
            Qt.WindowMaximizeButtonHint |
                                        ^
Found 9 errors in 5 files (checked 184 source files)
ERROR: InvocationError for command /home/terminal/code/qutebrowser/.tox/mypy/bin/python -m mypy qutebrowser tests (exited with code 1)
```

> It parses the output of git status --porcelain which does account for .gitignore - what does that command say on your system? I'd expect an empty output.

Yup, empty output.

Error for check-manifest says:
```
check-manifest run-test-pre: PYTHONHASHSEED='1847270991'
check-manifest run-test: commands[0] | /home/terminal/code/qutebrowser/.tox/check-manifest/bin/check-manifest --ignore 'qutebrowser/git-commit-id,qutebrowser/html/doc,qutebrowser/html/doc/*,*/__pycache__'
lists of files in version control and sdist do not match!
missing from VCS:
  qutebrowser/html/doc/img/cheatsheet-big.png
  qutebrowser/html/doc/img/cheatsheet-small.png
ERROR: InvocationError for command /home/terminal/code/qutebrowser/.tox/check-manifest/bin/check-manifest --ignore 'qutebrowser/git-commit-id,qutebrowser/html/doc,qutebrowser/html/doc/*,*/__pycache__' (exited with code 1)
```

**Reply by The-Compiler** [reply]
> My mistake, the package I was referencing is just called PyQtWebKit. Can't find anything about it in my repositories (on ArchLinux), but if it's not too relevant I might just skip it I suppose. App still runs fine.

On Archlinux, you can install `qt5-webkit` and then use `tox -e py` to run with Archlinux' PyQt/Qt install (rather than one installed from PyPI). However, unless you're working on something where you know you need to take special care for QtWebKit support, I probably wouldn't bother - QtWebKit integration is pretty much on life support (#4039).

> [mypy] Errors are: [...]

That looks like you're either not using the [PyQt5-stubs package](https://github.com/stlehmann/PyQt5-stubs), or using an outdated version - can you try running `tox -e mypy -r` to tell tox to recreate the environment? If you still see issues then, can you please show the full output including the version numbers at the top?

> Error for check-manifest says: [...]

Ah, my bad, I was mixing up `tox -e misc` with `check-manifest`. I suppose nobody reported this so far because it's usually run on CI with a clean checkout. Fixed in 23198a159 - thanks for the report!

**Comment by terminalnode** [comment]
Additionally, I noticed it wasn't running one of the tests (`py38-pyqt515-cov`) because I didn't have python38 installed. It's now running that with result:

`12 failed, 8238 passed, 356 skipped, 49 xfailed, 907 errors, 40 rerun in 1263.29s (0:21:03)`

**Reply by The-Compiler** [reply]
FWIW I usually explicitly use the environments I'm interested in, for example `tox -e py39-pyqt515`. Those errors are odd, though. Could you please show the test summary you get at the very end (one line per failed test, before the line you posted I think).
